### ISSUE TITLE
Export OpenTelemetry of Provider Execution

### ISSUE THREAD

**Issue by sempervictus** [opentofu/opentofu#2112]
### OpenTofu Version

```shell
OpenTofu v1.8.3
on linux_amd64
```

### The problem in your OpenTofu project

Using a fairly complex weave of providers for data and function which have various latency properties which may stack unfavorably. 

### Attempted Solutions

Skim of CLI help & docs for telemetry export options

### Proposal

Need to profile provider call execution preferably down to the underlying socket times communicating with the provider target but at least at the TF Read/Create/Update/etc level and export those metrics in standardized ways across running providers to inform users of which calls are taking the most time/resources.

**Comment by apparentlymart** [comment]
I believe that eventually including trace spans from provider plugins in the overall span for an operation is part of the intention of the proposed RFC https://github.com/opentofu/opentofu/pull/2028, though that is currently focused mainly on the general internal infrastructure work needed to make it viable to generate useful telemetry information _at all_ so we're likely still several steps away from exactly what this issue requests.

The exact details of how it would work are subject to change as that RFC and its successors evolve, but at a high level I'd imagined it working as follows:

1. Whatever is running OpenTofu CLI sets the relevant environment variables to tell OpenTofu where to send telemetry messages and, optionally, which parent trace to associate the generated spans with in case OpenTofu is itself operating as part of a broader system that already has its own telemetry.
2. When OpenTofu Core makes requests to provider plugins (using gRPC), it would use the existing OpenTelemetry adapters for propagating the trace context in the gRPC headers and encourage the plugin to send telemetry information to the same collector that OpenTofu CLI was told to use.
3. The plugin framework (running inside the plugin process) would then need to notice that extra metadata and propagate it into the plugin's own execution context, so that any telemetry spans generated by the provider would end up attached to the same trace that OpenTofu and its caller were using, collected by the same collector.

The existing RFC is mainly about finishing the work on the first step (already partially implemented) and making sure that the telemetry information can propagate _internally_ enough for the plugin client to be able to "see" it. Once that work is done, it should be a relatively small final step to add the relevant context-passing to the plugin client itself.

Of course, it'll ultimately be up to the provider plugin itself to actually make use of that information. The OpenTofu core implementation alone should provide enough information to know that, for example, OpenTofu sent the provider an `ApplyResourceChange` request and how long that request took in total, but anything beneath that span would be the provider's responsibility to generate and so not something we could directly control by changes made in this codebase.

Does what I've written above seem plausible, or did you have something else in mind? Thanks!

**Comment by sempervictus** [comment]
Thank you. My thought process re core is that even if the provider plugins do not export the data, their CRUD calls are all the same in terms of GRPC endpoints meaning that core can provide measurement of the call duration until some later version of the provider itself does not export internal metrics back to the core for relay/exposure to collectors.

**Comment by apparentlymart** [comment]
Thanks for the confirmation, @sempervictus.

I think in retrospect my direct reference to `ApplyResourceChange` (a specific operation in the plugin protocol) might have distracted from the general point I was trying to make, which is that OpenTofu Core _could_ generate telemetry information for calls made at the provider protocol level of abstraction regardless of what the provider does, and then the provider could optionally extend that with further details such as specific HTTP requests the provider made to satisfy the request.

So I think we are in agreement! The work in the proposed RFC is still required to make sure that the necessary context propagates properly throughout the OpenTofu codebase, since a lot of OpenTofu code long predates the Go convention of passing around `context.Context` containing telemetry metadata, but the ultimate goal of doing that was specifically to be able to export information like what this issue is requesting. Therefore I think we could consider this issue to be a retroactive additional feature request to further motivate https://github.com/opentofu/opentofu/pull/2028.

**Comment by apparentlymart** [comment]
Hi again, everyone!

The OpenTofu core team revisited this today. In the meantime, the previous OpenTelemetry proposal was retracted in favor of https://github.com/opentofu/opentofu/pull/2448, which we are still discussing but seems likely to happen in _some_ form.

That proposal focuses mainly on the cross-cutting concerns of how we'd integrate OpenTelemetry into OpenTofu in a way that allows it to be used as part of a larger distributed system with aggregated traces, and was intentionally light on proposing which specific actions we might represent in traces and exactly how they would be represented.

_This_ proposal is a good idea for a specific application of that work, which would allow those employing OpenTelemetry tracing to broadly understand which requests OpenTofu Core is making to provider plugins and how long they are taking, which can then be potentially used to optimize the typical runtime of changes made to a configuration.

With all of that said then: our conclusion was to accept this proposal but to leave it _on hold_ for the moment until we complete the discussion over in https://github.com/opentofu/opentofu/pull/2448. Assuming that RFC is accepted we'll need to first do some initial cross-cutting implementation work to plumb the OpenTelemetry libraries into OpenTofu more robustly, and then once that work is done we can revisit _this_ issue and define some specific details on how OpenTofu Core will report its interactions with providers using OpenTelemetry tracing concepts.

**Comment by remyleone** [comment]
Thanks for this work :) I got a couple of questions:
- What would be the requirements on the provider side to implement this feature? Is it compatible with the SDKv2 and/or the plugin framework?
- What level of details would be expected? As of today, the traces currently exported are very high level. Would a more function based tracing be possible? Would tracing of the gRPC requests be also supported?
- Is there specific implementation that would be nice to have on the provider side? For instance in how the API calls are happening, how the context is passed, how waits and retry are happening, etc... I'm wondering how much open telemetry will the provider need to know to make it happen.
- How can we trace interactions with the state?

**Comment by apparentlymart** [comment]
Our planning so far has been to:

- Get the internal trace metadata propagated to the gRPC clients within OpenTofu Core so that we can at least trace the outgoing requests.

    This effort is a little stalled right now because there are many different codepaths leading to gRPC requests and the design of the library we currently use for those means that we can't really turn it on for any of them until they are all ready. We have a bit of a "last mile problem" here getting from the current tracing detail to the next step. 

- Once the trace metadata can reach the gRPC clients we'll also arrange to propagate it in the gRPC headers so that the server side (the code inside the plugin) can emit its own connected traces if it wants to. 

    We don't currently control the main framework nor most providers, so this is as far as we can go with our current level of influence over the problem. But at least using the standard gRPC propagation formats means the server side integration can begin with preexisting code in the OpenTelemetry libraries.

This issue was about provider calls in particular and so I think other tracing like state interactions is out of scope here but we could discuss that more in one of the other telemetry issues, or in a new one if what your are thinking about isn't covered anywhere else yet. In particular though, our current intended scope is to focus on tracing higher-level operations that we expect would be meaningful to an OpenTofu user who is not deeply familiar with the internals. and so tracing at the individual function call level is not currently in scope. If we do choose to offer that later then I expect it would be disabled by default to avoid generating an overwhelming amount of detail that would be useful largely just to OpenTofu developers.
